build/cmake/Fuzzing.cmake

```diff
@@ -0,0 +1,131 @@
+option (ENABLE_FUZZING "Enable fuzzing using LLVM libFuzzer" OFF)
+
+if (ENABLE_FUZZING)
+    # This will add another sanitizer when we later include Sanitizers.cmake
+    list (APPEND MONGO_SANITIZE "fuzzer-no-link")
+endif ()
+
+include (ProcessorCount)
+ProcessorCount (_FUZZER_PARALLELISM)
+
+set (_FUZZERS_OUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/fuzzers")
+
+
+#[[
+    Generate an executable target that links and runs with LLVM libFuzzer.
+
+    Amongst the given source files there must be one definition of LLVMFuzzerTestOneInput.
+    Refer: https://www.llvm.org/docs/LibFuzzer.html#fuzz-target
+
+    This will define an executable with the given name, and all additional
+    arguments will be given as source files to that executable. This executable
+    will be linked with the '-fsanitize=fuzzer' command-line option.
+
+    This will additionally define a custom target "run-fuzzer-${name}," which,
+    when executed, will run the fuzzer executable with a set of pre-defined
+    libFuzzer command-line options.
+
+    The following target properties can be used to control the 'run-fuzzer'
+    target:
+
+    FUZZER_FORK (integer)
+        Set the number of parallel fuzzer tasks to run. The default is the
+        parallelism of the host plus four.
+
+    FUZZER_TIMEOUT (integer, seonds)
+        Set the maximum amount a single fuzzer task should run before the fuzzer
+        consideres it to be "stuck" and to generate a timeout report for the
+        given input.
+
+    FUZZER_LEN_CONTROL (integer, 1-100)
+        Set the len_control option for the libFuzzer run. Lower values tend to
+        generate larger inputs. Default is 50.
+
+    FUZZER_MAX_LEN (integer, bytes)
+        Set the maximum input size for a fuzzer input. The default is 4096.
+
+    FUZZER_ONLY_ASCII (boolean)
+        If TRUE, only valid ASCII will be given as fuzzer input.
+        The default is FALSE.
+
+    FUZZER_DICT (filepath)
+        Set to a filepath of a fuzzer dictionary.
+        Refer: https://www.llvm.org/docs/LibFuzzer.html#dictionaries
+        Default is to have no dictionary.
+
+    Fuzzer executables are written to to the <BUILD_DIR>/fuzzers directory.
+
+    This will unconditionally define the target and the custom target that
+    executes it, but it will be EXCLUDE_FROM_ALL=TRUE if the CMake setting
+    ENABLE_FUZZING is not true.
+]]
+function (mongoc_add_fuzzer name)
```

Comment:
Is the `run-fuzzer-bson-fuzz` target expected to complete quickly? I suspect I am doing something wrong or misunderstanding. But running `cmake --build cmake-build --target run-fuzzer-bson-fuzz` produces this output:

```
% cmake --build cmake-build --target run-fuzzer-bson-fuzz                
[0/1] cd /Users/kevin.albertson/review/mongo-c-driver-1000/cmake-build/fuzze...bertson/review/mongo-c-driver-1000/cmake-build/fuzzers/bson-fuzz.out//corpus

  Running fuzzer program : /Users/kevin.albertson/review/mongo-c-driver-1000/cmake-build/fuzzers/bson-fuzz.debug
     Corpus is stored in : /Users/kevin.albertson/review/mongo-c-driver-1000/cmake-build/fuzzers/bson-fuzz.out//corpus
  Crashes will appear in : /Users/kevin.albertson/review/mongo-c-driver-1000/cmake-build/fuzzers/bson-fuzz.out/

INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 2410153351
INFO: -fork=12: fuzzing in separate process(s)
INFO: -fork=12: 0 seed inputs, starting to fuzz in /var/folders/pv/p1jss0l97mq0ddr7rjbcbdt00000gp/T//libFuzzerTemp.FuzzWithFork16948.dir
#0: cov: 0 ft: 0 corp: 0 exec/s 0 oom/timeout/crash: 0/0/0 time: 0s job: 6 dft_time: 0
INFO: log from the inner process:
INFO: Seed: 2410369315
INFO:        0 files found in /var/folders/pv/p1jss0l97mq0ddr7rjbcbdt00000gp/T//libFuzzerTemp.FuzzWithFork16948.dir/C6
INFO: DataFlowTrace: reading from '/var/folders/pv/p1jss0l97mq0ddr7rjbcbdt00000gp/T//libFuzzerTemp.FuzzWithFork16948.dir/DFT'
INFO: A corpus is not provided, starting from an empty corpus
#2      INITED exec/s: 0 rss: 29Mb
INFO: 0/0 inputs touch the focus function
INFO: 0/0 inputs have the Data Flow Trace
ERROR: no interesting inputs were found. Is the code instrumented for coverage? Exiting.
INFO: exiting: 256 time: 0s
```

I expected the fuzzer to run for some time, and this completed instantly.

Reply:
That is abnormal. It means the execution is not generating any coverage data. I'll investigate when I get free time.